homeassistant/components/media_player/androidtv.py

```diff
@@ -0,0 +1,413 @@
+
+KNOWN_APPS = {
+    "dream": "Screensaver",
+    "kodi": "Kodi",
+    "netfilx": "Netflix",
+    "plex": "Plex",
+    "spotify": "Spotify",
+    "tvlauncher": "Homescreen",
+    "youtube": "Youtube"
+}
+
+ACTIONS = {
+    "back": "4",
+    "blue": "186",
+    "component1": "249",
+    "component2": "250",
+    "composite1": "247",
+    "composite2": "248",
+    "down": "20",
+    "end": "123",
+    "enter": "66",
+    "green": "184",
+    "hdmi1": "243",
+    "hdmi2": "244",
+    "hdmi3": "245",
+    "hdmi4": "246",
+    "home": "3",
+    "input": "178",
+    "left": "21",
+    "menu": "82",
+    "move_home": "122",
+    "mute": "164",
+    "pairing": "225",
+    "power": "26",
+    "resume": "224",
+    "right": "22",
+    "sat": "237",
+    "search": "84",
+    "settings": "176",
+    "sleep": "223",
+    "suspend": "276",
+    "sysdown": "281",
+    "sysleft": "282",
+    "sysright": "283",
+    "sysup": "280",
+    "text": "233",
+    "top": "122",
+    "up": "19",
+    "vga": "251",
+    "voldown": "25",
+    "volup": "24",
+    "yellow": "185"
+}
+
+DATA_KEY = '{}.androidtv'.format(DOMAIN)
+
+
+def setup_platform(hass, config, add_entities, discovery_info=None):
+    """Set up the androidtv platform."""
+
+    if DATA_KEY not in hass.data:
+        hass.data[DATA_KEY] = {}
+
+    from adb.client import Client as AdbClient
+    client = AdbClient(host="127.0.0.1", port=5037)
+
+    androidtvs = []
+    for host, device_config in config[CONF_DEVICES].items():
+        name = device_config[CONF_NAME]
+        port = str(device_config[CONF_PORT])
+        scan_interval = device_config[CONF_SCAN_INTERVAL]
+        uri = host + ':' + port
+
+        try:
+            ADBdev = client.device(uri)
+            if ADBdev is None:
+                _LOGGER.error(
+                    "Can't reach adb server, is it running ?")
+                raise PlatformNotReady
+
+            device = {'name': name,
+                      'ipaddr': host,
+                      'port': port,
+                      'scan_interval': scan_interval}
+            androidtv = AndroidTv(device)
+            androidtvs.append(androidtv)
+            hass.data[DATA_KEY][host] = androidtv
+        except RuntimeError:
+            _LOGGER.error(
+                "Can't reach adb server, is it running ?")
+            raise PlatformNotReady
+
+    add_entities(androidtvs, True)
+
+    def service(service):
+        """Dispatch service calls to target entities."""
+        params = {key: value for key, value in service.data.items()
+                  if key != ATTR_ENTITY_ID}
+
+        entity_ids = service.data.get(ATTR_ENTITY_ID)
+        if entity_ids:
+            target_devices = [dev for dev in hass.data[DATA_KEY].values()
+                              if dev.entity_id in entity_ids]
+        else:
+            target_devices = hass.data[DATA_KEY].values()
+
+        for target_device in target_devices:
+            if len(params.keys()) == 1:
```

Comment:
Why do we need this check? We should use a service voluptuous schema to validate service data.